rust/nasl-builtin-network/src/socket.rs

```diff
@@ -0,0 +1,731 @@
+    /// - length the number of bytes that you want to read at most. recv may return before length bytes have been read: as soon as at least one byte has been received, the timeout is lowered to 1 second. If no data is received during that time, the function returns the already read data; otherwise, if the full initial timeout has not been reached, a 1 second timeout is re-armed and the script tries to receive more data from the socket. This special feature was implemented to get a good compromise between reliability and speed when openvas-scanner talks to unknown or complex protocols. Two other optional named integer arguments can twist this behavior:
+    /// - min is the minimum number of data that must be read in case the “magic read function” is activated and the timeout is lowered. By default this is 0. It works together with length. More info https://lists.archive.carbon60.com/nessus/devel/13796
+    /// - timeout can be changed from the default.
+    fn recv(&self, r: &Register, _: &Context) -> Result<NaslValue, FunctionErrorKind> {
+        let socket = super::get_usize(r, "socket")?;
+        let len = super::get_usize(r, "length")?;
+        let _ = super::get_opt_int(r, "min");
+        let timeout = super::get_opt_int(r, "timeout");
+        let mut data = vec![0; len];
+
+        let mut ret = Ok(NaslValue::Null);
+
+        match self
+            .handles
+            .write()
+            .unwrap()
+            .handles
+            .get_mut(socket)
+            .ok_or(FunctionErrorKind::WrongArgument(format!(
+                "the given socket FD {socket} does not exist"
+            )))? {
+            NaslSocket::Tcp(conn) => {
+                let mut old = None;
+                if let Some(timeout) = timeout {
+                    old = conn.socket.read_timeout().unwrap();
+                    conn.socket
+                        .set_read_timeout(Some(Duration::from_secs(timeout as u64)))?;
+                }
+                // TODO: process min for magic read function"
+                match conn.socket.read(data.as_mut_slice()) {
+                    Ok(size) => {
+                        data.truncate(size);
+                        ret = Ok(NaslValue::Data(data));
+                    }
+                    Err(e) => ret = Err(e.into()),
+                };
+
+                if let Some(timeout) = old {
+                    conn.socket.set_read_timeout(Some(timeout))?;
+                }
+
+                ret
+            }
+            NaslSocket::Udp(conn) => {
+                let mut old = None;
+                if let Some(timeout) = timeout {
+                    old = conn.socket.read_timeout().unwrap();
+                    conn.socket
+                        .set_read_timeout(Some(Duration::from_secs(timeout as u64)))?;
+                }
+
+                let mut result = conn.socket.recv(data.as_mut_slice());
+
+                for _ in 0..5 {
```

Comment:
why 6 times?

Reply:
We resend the data up to 5 times. It is the same, as in the c implementation, except there we don't check if we can receive data after the last send. The c implementation also checks with `select` if there is data  available to receive, but I don't see any reason to do so, as it should be the same, as `recv` regarding the timeout. Maybe you have any idea?
https://github.com/greenbone/openvas-scanner/blob/main/nasl/nasl_socket.c#L790

Reply:
Since I had the same question as @nichtsfrei, I think the `5` at least deserves to be a named constant like `NUM_TIMES_TO_RESEND` or whatever.